### PR TITLE
Feats/backend

### PR DESCRIPTION
**Pull request by rafaljanwojcik**
### Proposed Changes:
- Added option for querying two indexes in the same Elasticsearch Document Store, one with faq embeddings, and one with the rulebook ( `rest_api/rest_api/pipeline/__init__.py` file)
- added option in `document_indexing/s3_storage.py` to download rulebook pdf file from s3 storage
- included appropriate changes in `document_indexing/populate_document_store.py` to populate two document stores with faq and rulebook docs
- included appropriate changes in Makefile (haystack-api needs quite some time to start running)
- modified .gitignore to not take into account downladed pdf rulebook
- modified system path definition of rest_api modules in `document_indexing/populate_document_store.py` and `document_indexing/s3_storage.py`, as it caused Import Errors in my code execution 

### How did you test it?
Didn't really implement any tests yet, tested everything locally in terminal, using Makefile, and testing provided queries to return expected outputs

### Notes for the reviewer
- many formatting changes in the code were done automatically by the default pre-commit hooks of haystack, also I had to comment out a few options of pre-commit-config YAML file, as original files from haystack library threw errors for them

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,16 +1,22 @@
-all: elastic_search_up index_documents give_indexing_some_time rest_api_up run_query
+all: elastic_search_up give_es_some_time populate_indexes rest_api_up give_api_some_time run_query run_faq_query run_rulebook_query
```
Comment (murielhol) [inline Makefile]: need to remove the run_query command

Comment (murielhol) [inline Makefile]:
```suggestion
all: elastic_search_up give_es_some_time populate_indexes rest_api_up give_api_some_time run_faq_query run_rulebook_query
```

File: rest_api/rest_api/pipeline/__init__.py [hunk 2]
```diff
@@ -10,6 +11,12 @@
 
 from rest_api.controller.utils import RequestLimiter
 
+import torch
```
Comment (murielhol) [inline rest_api/rest_api/pipeline/__init__.py]: duplicate

File: Makefile [hunk 3]
```diff
@@ -1,16 +1,22 @@
...
 
 rest_api_up:
 	docker-compose up -d haystack-api
 
-run_query:
-	 curl -H 'Content-Type: application/json' -H 'Accept: application/json' -d '{"query": "What is the objective of the game?"}' http://127.0.0.1:8000/query
+give_api_some_time:
+	sleep 120
+
+run_faq_query:
```
Comment (murielhol) [inline Makefile:18]: I am getting error when running this, is this running for you? If so, I will have a better look on why it does not work for me

File: document_indexing/s3_storage.py [hunk 4]
```diff
@@ -62,44 +64,39 @@ def upload_qa_pairs(self, qa_pairs: List[QuestionAnswerPair]):
...
-        with open(
-            f"s3://{self.bucket}/{game}.txt",
-            "r",
-            transport_params={"client": self.client},
-        ) as fin:
+        with open(f"s3://{self.bucket}/{game}.txt", "r", transport_params={"client": self.client}) as fin:
             serialized = fin.readlines()
         return [QuestionAnswerPair(**json.loads(s)) for s in serialized]
 
+    def _download_rulebook_from_s3(self, game: str, local_path: str) -> None:
```
Comment (murielhol) [inline document_indexing/s3_storage.py:90]: nice!

File: rest_api/rest_api/pipeline/__init__.py [hunk 5]
```diff
@@ -25,12 +32,70 @@ def setup_pipelines() -> Dict[str, Any]:
...
-    query_pipeline = Pipeline.load_from_yaml(Path(config.PIPELINE_YAML_PATH), pipeline_name=config.QUERY_PIPELINE_NAME)
+    # query_pipeline = Pipeline.load_from_yaml(Path(config.PIPELINE_YAML_PATH), pipeline_name=config.QUERY_PIPELINE_NAME)
+
+    # ------------------
+    document_store = ElasticsearchDocumentStore(host="host.docker.internal")
+    extractive_document_store = ElasticsearchDocumentStore(
+        host="host.docker.internal", index="rulebook", embedding_dim=768
+    )
+    faq_document_store = ElasticsearchDocumentStore(
+        host="host.docker.internal", index="faq", embedding_dim=384, similarity="cosine"
```
Comment (murielhol) [inline rest_api/rest_api/pipeline/__init__.py:42]: great solution! 🥳

File: Makefile (hunk 3 above)
Comment (murielhol) [reply Makefile:18]:
```
File "/usr/local/lib/python3.7/site-packages/rest_api/controller/search.py", line 78, in _process_request
    result = pipeline.run(query=request.query, params=params, debug=request.debug)
  File "/home/user/haystack/utils/reflection.py", line 32, in wrapper_invocation_counter
    return func(*args, **kwargs)
  File "/home/user/haystack/pipelines/base.py", line 477, in run
    self._validate_node_names_in_params(params=params)
  File "/home/user/haystack/pipelines/base.py", line 2105, in _validate_node_names_in_params
    f"No node(s) or global parameter(s) named {', '.join(invalid_keys)} found in pipeline."
ValueError: No node(s) or global parameter(s) named FaqRetriever, CustomClassifier found in pipeline.



```

Comment (rafaljanwojcik) [reply Makefile:18]: Hm, did you run `docker-compose build` in your repository directory, after pulling the branch? The docker images need to be re-built after every change in code

File: rest_api/rest_api/pipeline/__init__.py (hunk 5 above)
Comment (rafaljanwojcik) [reply rest_api/rest_api/pipeline/__init__.py:42]: Thanks! :)

File: rest_api/rest_api/pipeline/__init__.py [hunk 6]
```diff
@@ -25,12 +32,70 @@ def setup_pipelines() -> Dict[str, Any]:
     pipelines = {}
 
     # Load query pipeline
-    query_pipeline = Pipeline.load_from_yaml(Path(config.PIPELINE_YAML_PATH), pipeline_name=config.QUERY_PIPELINE_NAME)
+    # query_pipeline = Pipeline.load_from_yaml(Path(config.PIPELINE_YAML_PATH), pipeline_name=config.QUERY_PIPELINE_NAME)
+
+    # ------------------
+    document_store = ElasticsearchDocumentStore(host="host.docker.internal")
+    extractive_document_store = ElasticsearchDocumentStore(
+        host="host.docker.internal", index="rulebook", embedding_dim=768
```
Comment (timjones1) [inline rest_api/rest_api/pipeline/__init__.py:39]: Hey Rafal, I see you have set embedding_dim on the extractive document store. Did you try using an EmbeddingRetriever for the Extractive QA Pipeline as embedding_dims are not used by the BM25 Retriever, if so was the BM25 reader better in your view?

File: rest_api/rest_api/pipeline/__init__.py [hunk 7]
```diff
@@ -25,12 +32,70 @@ def setup_pipelines() -> Dict[str, Any]:
...
+        host="host.docker.internal", index="faq", embedding_dim=384, similarity="cosine"
+    )
+
+    extractive_reader_option = "deepset/roberta-base-squad2"
+    faq_retriever_option = "sentence-transformers/all-MiniLM-L6-v2"
+
+    faq_retriever = EmbeddingRetriever(
+        document_store=faq_document_store,
+        embedding_model=faq_retriever_option,
+        use_gpu=False,
```
Comment (timjones1) [inline rest_api/rest_api/pipeline/__init__.py]:
```suggestion
        use_gpu=torch.cuda.is_available(),
```
Hey Rafal, would it be better to check for GPU usage to speed up the update_embeddings call on the faq_document_store?

Comment (rafaljanwojcik) [reply rest_api/rest_api/pipeline/__init__.py]: Sure thing!

File: rest_api/rest_api/pipeline/__init__.py (hunk 6 above)
Comment (rafaljanwojcik) [reply rest_api/rest_api/pipeline/__init__.py:39]: I've set embedding_dim here to 768, but it's also the default of any DocumentStore in haystack, left it here for no particular reason tbh, apart from the fact that FARMReader's `deepset/roberta-base-squad2` has 768 dim embeddings, so maybe to emphasize that. Regarding the different retrievers - I didn't experiment with different architectures, used BM25, as in literature that's the best model (combination of BM25 + Dense reranking model), but for sure we can experiment later on with different sets of retrievers for extractive part

File: Makefile (hunk 3 above)
Comment (murielhol) [reply Makefile:18]: aiii haha wow, yes thanks! works

**Review by murielhol — APPROVED** [review]
great work!!

**Pull request merged**